### PR TITLE
GStreamer: Added new versions and conan v2 support

### PR DESCRIPTION
**Pull request by larshg**
Specify library name and version:  **1.14.5, 1.19.2 and 1.20.5**

Added several versions
Updated to support conan V2

superseeds and consolidates  #13537 and #13400

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by larshg** [comment]
Version >1.20 fails to build statically with the following error:
[1840/1843] Linking target gstreamer-full-1.0.dll
FAILED: gstreamer-full-1.0.dll
"link" @gstreamer-full-1.0.dll.rsp
Microsoft (R) Incremental Linker Version 14.34.31933.0
Copyright (C) Microsoft Corporation.  All rights reserved.

gstreamer-full-1.0.dll.rsp : fatal error LNK1170: line in command file contains 131071 or more characters

Seems like a meson / MSVC Linker issue.

Should static builds be disabled for those versions somehow?

File: recipes/gstreamer/config.yml [hunk 1]
```diff
@@ -1,13 +1,19 @@
...
     folder: all
-  "1.19.1":
+  "1.18.3":
     folder: all
-  "1.19.2":
+  "1.18.0":
+    folder: all
+  "1.16.2":
+    folder: all
+  "1.14.5":
```
Comment (prince-chrismc) [inline recipes/gstreamer/config.yml:18]: We usual refrain from adding older version, do you specifically need this one?

File: recipes/gstreamer/config.yml [hunk 2]
```diff
@@ -1,13 +1,19 @@
 versions:
-  "1.16.2":
+  "1.20.5":
     folder: all
-  "1.18.0":
+  "1.20.4":
```
Comment (prince-chrismc) [inline recipes/gstreamer/config.yml:4]: Since we are adding a more recent from the same minor, I'd suggest dropping this one since it's unlikely to be used

**Review by prince-chrismc — COMMENTED** [review]
This looks super good, just a few questions + minor suggestions

File: recipes/gstreamer/all/conanfile.py [hunk 3]
```diff
@@ -24,187 +35,236 @@ class GStreamerConan(ConanFile):
...
-        return "source_subfolder"
-
-    @property
-    def _build_subfolder(self):
-        return "build_subfolder"
 
-    def requirements(self):
-        self.requires("glib/2.72.0")
+    def configure(self):
+        if self.options.shared or self.settings.os == "Windows":
```
[the remark this answers is not in the document]
Comment (larshg) [reply recipes/gstreamer/all/conanfile.py:40]: I have just "merged" the condtions from _configure_ and _config_options_ - but they are called independently or?

File: recipes/gstreamer/all/conanfile.py [hunk 4]
```diff
@@ -24,187 +35,236 @@ class GStreamerConan(ConanFile):
...
+        if not self.dependencies.direct_host["glib"].options.shared and self.info.options.shared:
             # https://gitlab.freedesktop.org/gstreamer/gst-build/-/issues/133
-            raise ConanInvalidConfiguration("shared GStreamer cannot link to static GLib")
+            raise ConanInvalidConfiguration(
+                "shared GStreamer cannot link to static GLib")
 
     def build_requirements(self):
-        self.build_requires("meson/0.61.2")
-        self.build_requires("pkgconf/1.7.4")
+        if hasattr(self, "settings_build") and cross_building(self):
```
[the remark this answers is not in the document]
Comment (larshg) [reply recipes/gstreamer/all/conanfile.py:59]: I just followed the advice [here](https://github.com/conan-io/conan/issues/12768#issuecomment-1363304235) - atleast I hope I got it correct 😄

File: recipes/gstreamer/config.yml (hunk 1 above)
Comment (larshg) [reply recipes/gstreamer/config.yml:18]: Yes, since we are working with jetson pack 4.7.x, which has ubuntu 18.04 as base system, which is shipped with 1.14.5.
Evt. we will move to a newer version, but I don't know when.
I can remove it and then keep a local version if preferred.

**Comment by larshg** [comment]
When is 1.56 installed on the build servers? Should I try with 1.54?

**Comment by larshg** [comment]
Requires a fix added in 1.56.

**Comment by prince-chrismc** [comment]
> When is 1.56 installed on the build servers? Should I try with 1.54?

It will be a while, CCI is general 1 release behind. I'd suggest making it work with 1.54 or we'll have to be patient 😞

File: recipes/gstreamer/all/conanfile.py [hunk 5]
```diff
@@ -1,18 +1,29 @@
...
 class GStreamerConan(ConanFile):
     name = "gstreamer"
     description = "GStreamer is a development framework for creating applications like media players, video editors, streaming media broadcasters and so on"
-    topics = ("conan", "gstreamer", "multimedia", "video", "audio", "broadcasting", "framework", "media")
+    topics = ("multimedia", "video", "audio",
+              "broadcasting", "framework", "media")
     url = "https://github.com/conan-io/conan-center-index"
     homepage = "https://gstreamer.freedesktop.org/"
-    license = "GPL-2.0-only"
+    license = "LGPL-2.0-or-later"
```
Comment (prince-chrismc) [inline recipes/gstreamer/all/conanfile.py:26]: Why was this changed?

File: recipes/gstreamer/all/conanfile.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (prince-chrismc) [reply recipes/gstreamer/all/conanfile.py:40]: They are called independently. Most notably to the user before or after passing the command line input.

Our usual template give the best UX.

- if you set fpic on windows it's a hard error (since that just isnt a thing)
- If you pass fpic on linux but also shared it just disgarded since the value need to be true

you need to pass one option instead of worrying about both + this matches the internal logic of the helpers

File: recipes/gstreamer/all/conanfile.py [hunk 6]
```diff
@@ -24,187 +35,236 @@ class GStreamerConan(ConanFile):
...
 
-    def config_options(self):
-        if self.settings.os == 'Windows':
-            del self.options.fPIC
+    def requirements(self):
+        self.requires("glib/2.75.0")
 
     def validate(self):
-        if not self.options["glib"].shared and self.options.shared:
+        if not self.dependencies.direct_host["glib"].options.shared and self.options.shared:
```
Comment (prince-chrismc) [inline recipes/gstreamer/all/conanfile.py:53]: Ooo https://docs.conan.io/en/latest/reference/conanfile/dependencies.html?highlight=direct_host new stuff!

File: recipes/gstreamer/config.yml (hunk 1 above)
Comment (prince-chrismc) [reply recipes/gstreamer/config.yml:16]: No thats a good reason in my books :)

**Review by prince-chrismc — COMMENTED** [review]
Just one question about the license change :) otherwise it looks good + I tried to answer some of the previous stuff

File: recipes/gstreamer/all/conanfile.py (hunk 5 above)
Comment (larshg) [reply recipes/gstreamer/all/conanfile.py:26]: I did due to the other PR did it 😄 
But reading [this](https://gstreamer.freedesktop.org/documentation/frequently-asked-questions/licensing.html?gi-language=c)  it should be correct to change it to LGPL.

File: recipes/gstreamer/all/conanfile.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (larshg) [reply recipes/gstreamer/all/conanfile.py]: Not entirely sure how its notably to the user, if its removed in either calls, but I have reverted to having both _configure_ and _config_options_. I'll try read about the two.

[the remark this answers is not in the document]
Comment (larshg) [reply recipes/gstreamer/all/conanfile.py]: Ahh, I think I get it now. If they are removed in config_options, you get an error if you try set it in your conan call, since its no longer an option.
If removed in configure, you won't get an error even if you tried to set it - hence it can be more difficult to actually figure out if the value to try to set is actually used or not.

**Comment by larshg** [comment]
I think currently it fails with 1.54, due to this being [fixed](https://github.com/conan-io/conan/pull/12712), which is available from 1.56.
So I don't think we can fix it in the recipe - only by omitting the -pr:b profile, but I guess its not setable for the build on the CI?

**Comment by prince-chrismc** [comment]
c3i was update so lets give it a try

**Comment by larshg** [comment]
@prince-chrismc Something I can do about the timeout it mentions for gstreamer/1.19.1@macOS, Clang ?
